### PR TITLE
add unsafe gas/long limit checks to safety.

### PR DESCRIPTION
**Pull request by eFiniLan**
UNSAFE_DISABLE_DISENGAGE_ON_GAS (all cars)
UNSAFE_RAISE_LONGITUDINAL_LIMITS_TO_ISO_MAX (toyota)

### CODE REVIEW

File: board/safety/safety_honda.h [hunk 1]
```diff
@@ -196,8 +198,13 @@ static int honda_tx_hook(CAN_FIFOMailBox_TypeDef *to_send) {
 
   // disallow actuator commands if gas or brake (with vehicle moving) are pressed
   // and the the latching controls_allowed flag is True
-  int pedal_pressed = gas_pressed_prev || (gas_interceptor_prev > HONDA_GAS_INTERCEPTOR_THRESHOLD) ||
+  int pendal_pressed = 0;
+  if (unsafe_mode & UNSAFE_DISABLE_DISENGAGE_ON_GAS) {
```
Comment (geohot) [inline board/safety/safety_honda.h:202]: Use unsafe_allow_gas

Comment (geohot) [reply board/safety/safety_honda.h:202]: pendal_pressed is also misspelt. And can you rebase on master, sorry I introduced a misra violation.

File: board/safety/safety_honda.h [hunk 2]
```diff
@@ -196,8 +198,13 @@ static int honda_tx_hook(CAN_FIFOMailBox_TypeDef *to_send) {
 
   // disallow actuator commands if gas or brake (with vehicle moving) are pressed
   // and the the latching controls_allowed flag is True
-  int pedal_pressed = gas_pressed_prev || (gas_interceptor_prev > HONDA_GAS_INTERCEPTOR_THRESHOLD) ||
+  int pendal_pressed = 0;
```
Comment (geohot) [inline board/safety/safety_honda.h:201]: Can you refactor this to or in the gas stuff if (!unsafe_allow_gas)?

File: board/safety/safety_chrysler.h [hunk 3]
```diff
@@ -74,6 +74,8 @@ static int chrysler_rx_hook(CAN_FIFOMailBox_TypeDef *to_push) {
                                  chrysler_get_checksum, chrysler_compute_checksum,
                                  chrysler_get_counter);
 
+  bool unsafe_allow_gas = unsafe_mode & UNSAFE_DISABLE_DISENGAGE_ON_GAS;
```
Comment (geohot) [inline board/safety/safety_chrysler.h:77]: This is unused?

File: board/safety/safety_ford.h [hunk 4]
```diff
@@ -38,7 +39,7 @@ static int ford_rx_hook(CAN_FIFOMailBox_TypeDef *to_push) {
   // speed > 0
   if (addr == 0x165) {
     int brake_pressed = GET_BYTE(to_push, 0) & 0x20;
-    if (brake_pressed && (!brake_pressed_prev || ford_moving)) {
+    if (!unsafe_allow_gas && brake_pressed && (!brake_pressed_prev || ford_moving)) {
```
Comment (geohot) [inline board/safety/safety_ford.h:42]: This is for the brake, not the gas

File: board/safety/safety_toyota.h [hunk 5]
```diff
@@ -180,7 +185,12 @@ static int toyota_tx_hook(CAN_FIFOMailBox_TypeDef *to_send) {
           tx = 0;
         }
       }
-      bool violation = max_limit_check(desired_accel, TOYOTA_MAX_ACCEL, TOYOTA_MIN_ACCEL);
+      bool violation = false;
+      if (!(unsafe_mode & UNSAFE_RAISE_LONGITUDINAL_LIMITS_TO_ISO_MAX)) {
```
Comment (geohot) [inline board/safety/safety_toyota.h:189]: Perhaps inline ifs here on max_limit_check about which limit to use? No need to have a code branch.

File: board/safety/safety_honda.h [hunk 6]
```diff
@@ -194,16 +194,10 @@ static int honda_tx_hook(CAN_FIFOMailBox_TypeDef *to_send) {
...
-  // and the the latching controls_allowed flag is True
-  int pedal_pressed = gas_pressed_prev || (gas_interceptor_prev > HONDA_GAS_INTERCEPTOR_THRESHOLD) ||
-                      (brake_pressed_prev && honda_moving);
-  bool current_controls_allowed = controls_allowed && !(pedal_pressed);
-
   // BRAKE: safety check
   if ((addr == 0x1FA) && (bus == 0)) {
     honda_brake = (GET_BYTE(to_send, 0) << 2) + ((GET_BYTE(to_send, 1) >> 6) & 0x3);
-    if (!current_controls_allowed) {
+    if (!controls_allowed) {
```
Comment (geohot) [inline board/safety/safety_honda.h:200]: Why did this change?

File: board/safety/safety_gm.h [hunk 7]
```diff
@@ -136,16 +138,11 @@ static int gm_tx_hook(CAN_FIFOMailBox_TypeDef *to_send) {
...
-  // and the the latching controls_allowed flag is True
-  int pedal_pressed = gas_pressed_prev || (brake_pressed_prev && gm_moving);
-  bool current_controls_allowed = controls_allowed && !pedal_pressed;
-
   // BRAKE: safety check
   if (addr == 789) {
     int brake = ((GET_BYTE(to_send, 0) & 0xFU) << 8) + GET_BYTE(to_send, 1);
     brake = (0x1000 - brake) & 0xFFF;
-    if (!current_controls_allowed) {
+    if (!controls_allowed) {
```
Comment (geohot) [inline board/safety/safety_gm.h:145]: Why did this change?

**Comment by geohot** [comment]
I don't understand the current_ controls_allowed -> controls_allowed. If you want to make this change, I don't think this is the right pull request for it.

Also, can you write tests for the gas disengage like in #481 ?

**Comment by eFiniLan** [comment]
all good, test cases added!

File: board/safety/safety_ford.h [hunk 8]
```diff
@@ -72,7 +73,11 @@ static int ford_tx_hook(CAN_FIFOMailBox_TypeDef *to_send) {
 
   // disallow actuator commands if gas or brake (with vehicle moving) are pressed
   // and the the latching controls_allowed flag is True
-  int pedal_pressed = gas_pressed_prev || (brake_pressed_prev && ford_moving);
+  int pedal_pressed = brake_pressed_prev && ford_moving;
+  bool unsafe_allow_gas = unsafe_mode & UNSAFE_DISABLE_DISENGAGE_ON_GAS;
+  if (!unsafe_allow_gas) {
+    pedal_pressed = gas_pressed_prev;
```
Comment (geohot) [inline board/safety/safety_ford.h:79]: Where did the or go!?

**Comment by geohot** [comment]
Idk how the tests passed with this, there's no or on pedal_pressed.

Ideally, fix the tests to fail. Fix the bug. Then test should pass

File: board/safety/safety_gm.h [hunk 9]
```diff
@@ -138,7 +140,11 @@ static int gm_tx_hook(CAN_FIFOMailBox_TypeDef *to_send) {
 
   // disallow actuator commands if gas or brake (with vehicle moving) are pressed
   // and the the latching controls_allowed flag is True
-  int pedal_pressed = gas_pressed_prev || (brake_pressed_prev && gm_moving);
+  int pedal_pressed = brake_pressed_prev && gm_moving;
+  bool unsafe_allow_gas = unsafe_mode & UNSAFE_DISABLE_DISENGAGE_ON_GAS;
+  if (!unsafe_allow_gas) {
+    pedal_pressed = gas_pressed_prev;
```
Comment (geohot) [inline board/safety/safety_gm.h:146]: There's still no or on this!

Comment (eFiniLan) [reply board/safety/safety_gm.h:146]: lol just went out, will let you know once I'm done 👍

**Comment by eFiniLan** [comment]
"or" is added and 2 additional test cases for tx_hook when pedal is pressed.

File: board/safety/safety_toyota.h [hunk 10]
```diff
@@ -180,7 +185,9 @@ static int toyota_tx_hook(CAN_FIFOMailBox_TypeDef *to_send) {
           tx = 0;
         }
       }
-      bool violation = max_limit_check(desired_accel, TOYOTA_MAX_ACCEL, TOYOTA_MIN_ACCEL);
+      bool violation = !(unsafe_mode & UNSAFE_RAISE_LONGITUDINAL_LIMITS_TO_ISO_MAX)?
```
Comment (geohot) [inline board/safety/safety_toyota.h:188]: Please add parens here, I don't know the order of operations

Comment (dragonpilot) [reply board/safety/safety_toyota.h]: flipped!

**Comment by geohot** [comment]
Okay, LGTM. Let me know when you've testing in dragonpilot that this has the desired behavior

**Comment by eFiniLan** [comment]
cool! thanks for taking some time reviewing the PR. :)
Have tested this on toyota/lexus/honda and are working as expected.

Also noticed that the safety_toyota is not detecting gas signal on hybrid car correctly (0x245), will create a separate PR for it.

cheers!

**Pull request merged**